File: plugins/plugin-commands/rust/src/actions/commands_list.rs
use async_trait::async_trait;
use serde_json::Value;

use crate::registry::CommandRegistry;
use crate::{Action, ActionExample, ActionResult};

pub struct CommandsListAction;

#[async_trait]
impl Action for CommandsListAction {
    fn name(&self) -> &str {
        "COMMANDS_LIST_COMMAND"
    }

    fn similes(&self) -> Vec<&str> {
        vec!["/commands", "/cmds"]
    }

    fn description(&self) -> &str {
        "List all registered commands with their aliases. Only activates for /commands or /cmds."
    }

    async fn validate(&self, message: &Value, _state: &Value) -> bool {
        let text = message
            .get("content")
            .and_then(|c| c.get("text"))
            .and_then(|t| t.as_str())
            .unwrap_or("");
        if let Some(parsed) = crate::parser::parse_command(text) {
            matches!(parsed.name.as_str(), "commands" | "cmds")
        } else {
            false
        }
    }

    async fn handler(
        &self,
        _message: &Value,
        _state: &Value,
        registry: Option<&CommandRegistry>,
    ) -> ActionResult {
        let Some(reg) = registry else {
            return ActionResult {
                success: false,
                text: "Command registry is not available.".to_string(),
                data: None,
                error: Some("missing_registry".to_string()),
            };
        };

        let all = reg.list_all();
        let count = all.len();
        let mut lines = vec![format!("**Commands ({}):**", count), String::new()];

        for cmd in &all {
            let aliases = if cmd.aliases.is_empty() {
                format!("/{}", cmd.name)
            } else {
                let alias_strs: Vec<String> =
                    cmd.aliases.iter().map(|a| format!("/{}", a)).collect();
                format!("/{}, {}", cmd.name, alias_strs.join(", "))
            };
            let hidden_note = if cmd.hidden { " [hidden]" } else { "" };
            lines.push(format!(
                "  **{}**: {}{}",
                cmd.name, aliases, hidden_note
            ));
        }

        let text = lines.join("\n");
        ActionResult {
            success: true,
            text,
            data: Some(serde_json::json!({ "commandCount": count })),
            error: None,
        }
    }

    fn examples(&self) -> Vec<ActionExample> {
        vec![ActionExample {
            user_message: "/commands".to_string(),
            agent_response: "**Commands (5):**\n\n  **help**: /help, /h, /?...\n  **status**: /status, /s...".to_string(),
        }]
    }
}

File: plugins/plugin-commands/typescript/src/actions/commands-list.ts
/**
 * Commands list action
 */

import type {
  Action,
  ActionExample,
  IAgentRuntime,
  Memory,
} from "@elizaos/core";
import { getEnabledCommands } from "../registry";
import { detectCommand } from "../parser";

export const commandsListAction: Action = {
  name: "COMMANDS_LIST",
  description:
    "List all available commands with their aliases. Only activates for /commands or /cmds slash commands.",
  // Only slash-command similes to avoid matching natural language
  similes: ["/commands", "/cmds"],

  async validate(runtime: IAgentRuntime, message: Memory): Promise<boolean> {
    const text = message.content?.text ?? "";
    // Strict: only activate for slash commands, never natural language
    const detection = detectCommand(text);
    return detection.isCommand && detection.command?.key === "commands";
  },

  async handler(
    runtime: IAgentRuntime,
    message: Memory,
    state,
    options,
    callback,
  ) {
    const commands = getEnabledCommands();
    const lines: string[] = [`**Commands (${commands.length}):**\n`];

    for (const cmd of commands) {
      const aliases = cmd.textAliases.join(", ");
      const authNote = cmd.requiresAuth ? " [auth]" : "";
      const elevatedNote = cmd.requiresElevated ? " [elevated]" : "";
      lines.push(`• **${cmd.key}**: ${aliases}${authNote}${elevatedNote}`);
    }

    const replyText = lines.join("\n");
    await callback?.({ text: replyText });

    return {
      success: true,
      text: replyText,
      data: { commandCount: commands.length },
    };
  },

  examples: [
    [
      { name: "user", content: { text: "/commands" } },
      {
        name: "assistant",
        content: {
          text: "**Commands (15):**\n\n• **help**: /help, /h, /?\n• **status**: /status, /s...",
        },
      },
    ],
  ] as ActionExample[][],
};
